### ISSUE TITLE
[PopUpMenuButton] [Proposal] Add the ability to customise Modal Barrier Behaviour

### ISSUE THREAD

**Issue by nosmirck** [flutter/flutter#98211]
## Introduction

As discussed in #90223 the PopUpMenuButton (and I also found that DropdownButton and DropdownButtonFormField) doesn't close when scrolling. The Material behavior seems to describe that any gesture would close it, but it's not. This is being fixed by this PR: #98191 

However, I also mentioned about a proposal to allow drag/scroll gestures to be passed down to the layer behind so the content behind the modal can be scrolled (possibly with the popUp content too).

## Proposal

Add some kind of behavior (given by an enum value) to the widgets that use a modal barrier. The possible behaviors are:
- closeOnAnyTap: this is the default Material behavior (after fixing the bug that the modal doesn't dismiss with drag/scroll) for everything (mobile, desktop, web)
- passThrough: allows drag gestures (not taps, taps close the modal) to be passed down to the widgets behind it, this allows some Desktop/Web behavior for pop up menus and dropdowns (an example is right here on GitHub where you can click the top right kebab menu ( **. . .** ) and then scroll the page (web desktop only, on mobile it closes immediately with any interaction) you'll notice that the menu keeps open, and it keeps scrolling alongside the origin of the menu with the page itself.

## Use case

When building large forms on Flutter web, this is basically a need, users might need to scroll to get some other fields visible while they have the popUp open to make a decision on what to pick.

When making desktop/web apps that have menus (like File, Edit, View, Insert, Format, etc for Docs on Google Docs) you can open a specific menu and then scroll the content of the page, some users do this on the same motivation as the previous one, to check some content before making a choice of the menu options (they just want to have the options visible while looking at the content they are working on)

## Proposal

The widgets using the modal barrier can look now something like this:

```dart
PopupMenuButton<int>(
  icon: const Icon(Icons.view_list_rounded),
  //closes the modal with any gesture (as soon as the cursor taps/clicks down) this is the default behavior
  //modalBehavior: ModalBehavior.closeOnAnyTap,
  modalBehavior: ModalBehavior.passThrough,  //lets the scrolling/dragging gestures be passed down
  onSelected: (_) {},
  itemBuilder: (context) => List.generate(
    100,
    (i) => PopupMenuItem(
      child: Text('Pop Up Menu Item $i'),
      value: i,
    ),
  ),
),
```

P.S: Don't take the naming I gave as solid, I'm the worst at naming things, so, make it something that makes sense in the context of flutter/menus/popups/modals/whatever you think is best :)

**Comment by dkwingsmt** [comment]
Shall we rename the issue to be "Add an option for gesture to pass through modal barrier", which I think more tightly summarize what you propose? Your current title suggests that there are other modes of "behavior".

Also, I think if we follow the behavior of Github, then your description is not correct. For example, while tap does close the drop down menu, the same tap is also passed to the widget below (try to click another button with the drop down menu open).

I think the behavior should be:
* All gestures should pass through (including tapping, scrolling, hovering, etc.)
* Pointer down events should close the menu.

I wonder if we can already implement it now since we have `TapRegion`, though we can make it more accessible for the drop down menu.

**Comment by nosmirck** [comment]
@dkwingsmt 
> Also, I think if we follow the behavior of Github, then your description is not correct.

Try on Web (this is mainly for Web/Desktop) open the little kebab menu at the top right of this response (...), you should see Copy Link, Quote Reply, etc. if you leave that menu open and scroll the page (use the mouse wheel) the menu remains open and the pop up remains visible, following the position of the scroll. If you have it open and click any other button it will not execute the button but close the menu. I'm trying this right now as I write this reply on Chrome on my Mac, but I've tested this on Windows and it's the same behaviour (for Chrome at least). On Mobile we have a totally different story, although, having the option to passthrough the gestures would be awesome too (for the same reason I opened this ticket, we want to be able to scroll large forms as users are filling them)

> All gestures should pass through (including tapping, scrolling, hovering, etc.)

Not necessarily, sometimes we want gestures to be captured by the barrier so they are not passed down.

> Pointer down events should close the menu.

If you scroll that counts as a pointer down, which will close the popup, defeating the purpose of this feature.

**Comment by dkwingsmt** [comment]
> If you have it open and click any other button it will not execute the button but close the menu. 

Ok that's interesting. First of all, I confirm your description. But also if (on desktop) you scroll to the top of the page, and click the `+ ` button (named _Create new..._), it also pops up a menu, and if you click other buttons in the page, the menu open and the click also goes through. I guess Github is inconsistent by itself, and also there are different extends of the pass through behavior.

**Comment by nosmirck** [comment]
Yes, and I've seen other types of behaviours, like on steam, if you open a popup menu on a page content and scroll up/down, the menu remains in the same position relative to the screen instead of to the parent button.

There are many things that can be done, but for a starter, we should have a way to optionally pass down a gesture (desktop and mobile).

One very useful use case is for instance the menus on top of apps like Word, if you open one of the top menus (like file or edit) it can remain open while you scroll the content behind it, for instance, to disocclude content you're editing occluded by the menu opening.

**Comment by helpforsoftware** [comment]
please fix this issue . year 2025
